### PR TITLE
GH-1845: Extended the LSP with the semantic highlighting capabilities.

### LINKED ISSUE #1845: Semantic Coloring
This would be useful for embedded dsls. Thoughts?

### PR DESCRIPTION
**Pull request by kittaakos**
Closes #1845 

Signed-off-by: Akos Kitta <kittaakos@typefox.io>

### CODE REVIEW

File: packages/core/src/common/disposable.ts [hunk 1]
```diff
@@ -32,9 +32,16 @@ export namespace Disposable {
 }
 
 export class DisposableCollection implements Disposable {
+
     protected readonly disposables: Disposable[] = [];
     protected readonly onDisposeEmitter = new Emitter<void>();
 
+    constructor(toDispose?: Disposable | Disposable[]) {
```
Comment (akosyakov) [inline packages/core/src/common/disposable.ts]: Nice idea! How about using varargs?

```ts
constructor(...disposables: Disposable[]) {
    this.pushAll(disposables);
}
```

File: packages/editor/src/browser/semantic-highlight/semantic-highlighting-service.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+export const SemanticHighlightingService = Symbol('SemanticHighlightingService');
+export interface SemanticHighlightingService extends Disposable {
+
+    decorate(uri: URI, ranges: SemanticHighlightingRange[]): Promise<void>;
+
+}
+
+@injectable()
+export class NoopSemanticHighlightingService implements SemanticHighlightingService {
```
Comment (akosyakov) [inline packages/editor/src/browser/semantic-highlight/semantic-highlighting-service.ts]: How about turning `SemanticHighlightingService` into a class and use it as no-op implementation to get rid of additional symbol, inteface and bindings? Similary to `QuickOpenService`.

File: package.json [hunk 3]
```diff
@@ -23,6 +23,7 @@
     "chai-string": "^1.4.0",
     "concurrently": "^3.5.0",
     "electron-mocha": "~3.5.0",
+    "esm": "^3.0.72",
```
Comment (kittaakos) [inline package.json]: TODO: remove this.

File: packages/java/src/browser/monaco-contribution/java-textmate-contribution.ts [hunk 4]
```diff
@@ -15,33 +15,42 @@
...
 
 import { injectable } from 'inversify';
-import { JAVA_LANGUAGE_ID } from '../../common';
+// import { JAVA_LANGUAGE_ID } from '../../common';
 import { LanguageGrammarDefinitionContribution, TextmateRegistry } from '@theia/monaco/lib/browser/textmate';
 
 @injectable()
 export class JavaTextmateContribution implements LanguageGrammarDefinitionContribution {
 
+    // TODO: restore this!
```
Comment (kittaakos) [inline packages/java/src/browser/monaco-contribution/java-textmate-contribution.ts]: TODO!!!

File: packages/monaco/src/browser/monaco-semantic-highlighting-service.ts [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        }
+        return undefined;
+    }
+
+    /**
+     * Returns with a range between `from` and `to`: `[from, from + 1, ..., to - 1, to]`.
+     * - `from` can be greater than `to`. If so, the arguments will be swapped internally and you still get back `[to, to + 1, ..., from -1, from]`
+     * - If `from` and `to` are the same, returns with a single element range: `[from]`.
+     */
+    protected rangeOf(from: number, to: number): number[] {
```
Comment (kittaakos) [inline packages/monaco/src/browser/monaco-semantic-highlighting-service.ts]: TODO: remove this.

File: packages/monaco/src/browser/monaco-semantic-highlighting-service.ts [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        return {};
+    }
+
+    protected tokenTheme(): monaco.services.TokenTheme {
+        return StaticServices.standaloneThemeService.get().getTheme().tokenTheme;
+    }
+
+}
+
+// /**
```
Comment (kittaakos) [inline packages/monaco/src/browser/monaco-semantic-highlighting-service.ts]: TODO: clean this up.

File: packages/java/scripts/download-jdt-ls.js [hunk 7]
```diff
@@ -26,9 +26,10 @@ const path = require('path');
 const packageJson = require('../package.json');
 const shared = require('./shared');
 const packagePath = path.join(__dirname, "..");
-const serverPath = packageJson['ls.download.path'] || '/che/che-ls-jdt/snapshots/che-jdt-language-server-latest.tar.gz';
-const downloadURI = packageJson['ls.download.base'] || 'https://www.eclipse.org/downloads/download.php?file=';
-const archiveUri = downloadURI + serverPath + '&r=1';
+// TODO revert this!
+const serverPath = 'https://github.com/kittaakos/eclipse.jdt.ls-gh-715/raw/master/jdt-language-server-latest.tar.gz';//packageJson['ls.download.path'] || '/che/che-ls-jdt/snapshots/che-jdt-language-server-latest.tar.gz';
```
Comment (kittaakos) [inline packages/java/scripts/download-jdt-ls.js]: TODO: It is a hack. Revert this.

File: packages/editor/src/browser/semantic-highlight/semantic-highlighting-service.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable } from 'inversify';
+import { Position, Range } from 'vscode-languageserver-types';
+import URI from '@theia/core/lib/common/uri';
+import { Disposable } from '@theia/core/lib/common/disposable';
+
+@injectable()
+export class SemanticHighlightingService implements Disposable {
```
Comment (svenefftinge) [inline packages/editor/src/browser/semantic-highlight/semantic-highlighting-service.ts]: We should add some comment, explaining why this impl is entirely empty. It's ok to have a pointer to the monaco implementation in it.

File: packages/monaco/src/browser/monaco-semantic-highlighting-service.ts [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+        const affectedLines = Array.from(new Set(ranges.map(r => [r.start.line, r.end.line]).reduce((prev, curr) => prev.concat(curr), [])));
+        return affectedLines
+            .map(line => editor.getLinesDecorations(line, line))
+            .reduce((prev, curr) => prev.concat(curr), [])
+            .map(decoration => decoration.id)
+            .filter(id => ids.has(id));
+    }
+
+    /**
+     * We do not get delta notification from the LS if lines were deleted and new semantic highlighting positions were introduced.
```
Comment (svenefftinge) [inline packages/monaco/src/browser/monaco-semantic-highlighting-service.ts]: I remember we talked about it, but I forget why that is. I think it should be the responsibility of the LS to send updates in that case, too.

File: packages/editor/src/browser/editor.ts [hunk 10]
```diff
@@ -89,17 +89,40 @@ export interface TextEditor extends Disposable, TextEditorSelection, Navigatable
      */
     deltaDecorations(params: DeltaDecorationParams): string[];
 
+    /**
+     * Get the range associated with a decoration.
+     * @param id The decoration id.
+     * @return The decoration range or `undefined` if the decoration was not found.
+     */
+    getDecorationRange(id: string): Range | undefined;
```
Comment (svenefftinge) [inline packages/editor/src/browser/editor.ts]: I couldn't find any clients for this.

File: packages/editor/src/browser/semantic-highlight/semantic-highlighting-service.ts (hunk 8 above)
Comment (kittaakos) [reply packages/editor/src/browser/semantic-highlight/semantic-highlighting-service.ts]: I can move the entire implementation from `@theia/monaco` to `@theia/editor` if I can expose the `id` property for the editor decoration.

The API for the `EditorDecoration` lacks this information.

File: packages/java/scripts/download-jdt-ls.js (hunk 7 above)
Comment (kittaakos) [reply packages/java/scripts/download-jdt-ls.js]: https://github.com/theia-ide/theia/pull/2519

**Comment by kittaakos** [comment]
@tsmaeder, what is the difference between the https://www.eclipse.org/downloads/download.php?file=/che/che-ls-jdt/snapshots/che-jdt-language-server-latest.tar.gz and 
http://download.eclipse.org/jdtls/snapshots/jdt-language-server-latest.tar.gz LSs? Thanks!

**Comment by kittaakos** [comment]
Bumping the test timeout fixed the issue. I do one more squash and build, then I merge this.

File: packages/java/src/node/java-cli-contribution.ts [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable } from 'inversify';
+import { Argv, Arguments } from 'yargs';
+import { CliContribution } from '@theia/core/src/node/cli';
```
Comment (akosyakov) [inline packages/java/src/node/java-cli-contribution.ts]: `src` -> `lib`

**Pull request merged**